KAFKA-17057: Add RETRY option to ProductionExceptionHanlder

mjsax opened the pull request:
Implements KIP-1065

mjsax committed:
KAFKA-17057: Add RETRY option to ProductionExceptionHanlder

Implements KIP-1065

aliehsaeedii commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:
```diff
@@ -393,39 +393,41 @@ private void recordSendError(final String topic,
+                return;
+            }
+
+            if (productionException instanceof RetriableException && handlerResponse == ProductionExceptionHandlerResponse.RETRY) {
```
Thanks Matthias. Much better/more general than `isRetriable()`.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/errors/ProductionExceptionHandler.java:
```diff
@@ -85,18 +85,37 @@ default ProductionExceptionHandlerResponse handleSerializationException(final Er
+        /** Continue processing.
+         *
+         * <p> For this case, output records which could not be written successfully are lost.
+         * Use this option only, if your can tolerate data loss.
```
```suggestion
         * Use this option only if you can tolerate data loss.
```

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:
```diff
@@ -393,39 +393,41 @@ private void recordSendError(final String topic,
-            if (isRetriable(productionException)) {
+            final DefaultErrorHandlerContext errorHandlerContext = new DefaultErrorHandlerContext(
+                null, // only required to pass for DeserializationExceptionHandler
+                context.recordContext().topic(),
```
Do we need a check that the `recordContext()` itself is not `null` first?

mjsax replied:
Guess you are right -> https://issues.apache.org/jira/browse/KAFKA-17527 (working on a PR for a fix already, and can update this PR later for it...)

mjsax replied:
Other PR was merged. Resolved via rebasing.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/integration/SwallowUnknownTopicErrorIntegrationTest.java:
```diff
@@ -140,40 +166,27 @@ private void closeApplication(final Properties streamsConfiguration) throws Exce
-
-        return true;
+        closeApplication(streamsConfiguration);
+        assertEquals(kafkaStreams.state(), State.NOT_RUNNING);
```
Just curious - do we need to assert the application is not running at the end of test?  I don't recall us doing this in other integration tests.

mjsax replied:
Not sure yet -- the idea was to check if there was not crash, and final state is not ERROR -- but I think I want to change this test a little bit more anyway... (not 100% happy with the test yet)

mjsax replied:
After testing a few more things, the test is actually ok -- when I change the test handler to return RETRY or FAIL (instead of CONTINUE), the test does on pass. Will remove this check; it's not necessary.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java:
```diff
@@ -1347,15 +1348,40 @@ public void shouldThrowStreamsExceptionOnUnknownTopicOrPartitionExceptionWithDef
+    }
+
+    @Test
+    public void shouldThrowStreamsExceptionOnUnknownTopicOrPartitionExceptionExceptionHandlerReturnsFail() {
```
nit - The name `shouldThrowStreamsExceptionOnUnknownTopicOrPartitionExceptionExceptionHandlerReturnsFail` is bit hard to grok what about `shouldThrowStreamsExceptionOnUnknownTopicOrPartitionExceptionWhenExceptionHandlerReturnsFail`

totally optional

bbejeck reviewed (commented):
Overall it lgtm, I've left a couple of comments

mjsax committed:
github comments and rebasing

mjsax commented:
Build failure resolved by rebasing. Also addressed review comments expect the one about `null` check -- we should merge the other PR first, because the other PR need to be cherry-picked to `3.9`, too, and rebase this PR afterwards.

**THUS: DON'T MERGE THIS PR** before we could merge the other one and rebase this one.

mjsax committed:
checkstyle

mjsax commented:
Rebased and address all comments. Hope I got it right; please take another look.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/errors/ProductionExceptionHandler.java:
```diff
@@ -85,18 +85,37 @@ default ProductionExceptionHandlerResponse handleSerializationException(final Er
+         *
+         * <p> Kafka Streams will raise an exception and the {@code StreamsThread} will fail.
+         * No offsets (for {@link org.apache.kafka.streams.StreamsConfig#AT_LEAST_ONCE at-least-once}) or transactions
+         * (for {@link org.apache.kafka.streams.StreamsConfig#EXACTLY_ONCE_V2 }exactly-once) will be committed.
```
```suggestion
         * (for {@link org.apache.kafka.streams.StreamsConfig#EXACTLY_ONCE_V2 exactly-once}) will be committed.
```

mjsax committed:
Update streams/src/main/java/org/apache/kafka/streams/errors/ProductionExceptionHandler.java

bbejeck reviewed (approved):
LGTM
